pallets/pallet-governance/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+				if let Some(Ok(call)) = Proposals::<T>::take(proposal)
+					.map(|encoded| <T as Config>::RuntimeCall::decode(&mut &(*encoded)))
+				{
+					if approved {
+						Self::deposit_event(Event::<T>::ProposalPassed {
+							id: proposal,
+							call: call.clone(),
+							result: Self::dispatch_governance_call(call),
+						});
+					} else {
+						Self::deposit_event(Event::<T>::ProposalRejected {
+							id: proposal,
+							reason: RejectReason::ByVoting,
+						});
+					}
+				}
+				// Deleted the related data in Votes.
+				Votes::<T>::remove(proposal);
+			});
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Registered a proposal.
+		///
+		/// Require a member from council.
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::initiate_proposal())]
+		pub fn initiate_proposal(
+			origin: OriginFor<T>,
+			call: <T as Config>::RuntimeCall,
+		) -> DispatchResult {
+			let id = ensure_signed(origin)?;
+			ensure!(CurrentAuthorities::<T>::get().contains(&id), Error::<T>::InvalidAuthority);
+			let proposal_id = Self::next_proposal_id();
+			let expired_block =
+				T::BlockNumberProvider::current_block_number() + T::ExpiryPeriod::get();
+
+			// add call to storage
+			Proposals::<T>::insert(proposal_id, call.encode());
+
+			// add proposal id to expiry.
+			// Get the BTreeSet for the block number or create a new one if it doesn't exist
+			let mut expiry_set = Expiry::<T>::get(expired_block).unwrap_or_default();
+
+			// Insert the proposal id into the set
+			expiry_set.insert(proposal_id);
+
+			// Store the updated set back into storage
+			Expiry::<T>::insert(expired_block, expiry_set);
+
+			// add vote to Votes
+			let mut set = BTreeSet::new();
+			set.insert(id.clone());
+			Votes::<T>::insert(proposal_id, CastedVotes { yays: set, nays: Default::default() });
+
+			// Emit Event: ProposalRegistered
+			Self::deposit_event(Event::<T>::ProposalRegistered { who: id, call });
+
+			Ok(())
+		}
+
+		// Requires Root origin - sets the current authorities
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::vote())]
+		pub fn vote(origin: OriginFor<T>, proposal: ProposalId, approve: bool) -> DispatchResult {
+			// check if the vote can be casted
+			// is member of the council
+			let id = ensure_signed(origin)?;
+			ensure!(CurrentAuthorities::<T>::get().contains(&id), Error::<T>::InvalidAuthority);
+
+			// Proposal is valid
+			ensure!(Proposals::<T>::get(proposal).is_some(), Error::<T>::InvalidProposalId);
+
+			if let Some(mut casted) = Votes::<T>::get(proposal) {
+				// has not voted
+				ensure!(
+					!casted.yays.contains(&id) && !casted.nays.contains(&id),
+					Error::<T>::AlreadyVoted
+				);
+
+				// Cast the vote
+				if approve {
+					casted.yays.insert(id.clone());
+				} else {
+					casted.nays.insert(id.clone());
+				}
+
+				// check if the proposal is ready to be resolved.
+				let count_yays = casted.yays.len();
```

Comment:
This can also be an internal function in the CastedVotes struct.

```
/// Checks if there are enough votes to resolve a proposal. None if not enough votes.
/// If yes, then the result is returned as a bool (true = accept, false = reject)
fn can_resolve(&self, threshold: Percent) -> Option<bool>;
```